caikit/runtime/service_generation/data_stream_source.py

```diff
@@ -52,124 +57,74 @@
-
-        # If jsonfile, attempt to read the file and pull in the data from there
-        if set_field == "file":
-            return self._create_data_stream_from_file(fname=self.file.filename)
-
-        # If list of files, attempt to read all files and combine datastreams from all
-        if set_field == "listoffiles":
-            # combined list of data streams that we will chain and send back
-            data_stream_list = []
-            for fname in self.listoffiles.files:
-                data_stream_list.append(self._create_data_stream_from_file(fname=fname))
-
-            return DataStream.chain(data_stream_list).flatten()
-
-        # If directory, attempt to read an element from each file in the dir
-        if set_field == "directory":
-            dirname = self.directory.dirname
-            full_dirname = self._get_resolved_source_path(dirname)
-            extension = self.directory.extension or "json"
-            if not dirname or not os.path.isdir(full_dirname):
-                raise CaikitRuntimeException(
-                    grpc.StatusCode.INVALID_ARGUMENT,
-                    f"Invalid {extension} directory source file: {full_dirname}",
-                )
-            files_with_ext = list(glob(os.path.join(full_dirname, "*." + extension)))
-            # make sure at least 1 file with the given extension exists
-            if len(files_with_ext) == 0:
-                raise CaikitRuntimeException(
-                    grpc.StatusCode.INVALID_ARGUMENT,
-                    f"directory {dirname} contains no source files with extension {extension}",
-                )
-            if extension == "json":
-                return DataStream.from_json_collection(full_dirname, extension).map(
-                    self._to_element_type
-                )
-            if extension == "csv":
-                return DataStream.from_csv_collection(full_dirname).map(
-                    self._to_element_type
-                )
-            if extension == "jsonl":
-                return DataStream.from_jsonl_collection(full_dirname).map(
-                    self._to_element_type
-                )
-            raise CaikitRuntimeException(
-                grpc.StatusCode.INVALID_ARGUMENT,
-                f"Extension not supported! {extension}",
-            )
+    ## Shared Impl ##
+
+    @staticmethod
+    def _to_element_type(element_type: type, raw_element: Any) -> Any:
+        if issubclass(element_type, DataBase):
+            # To allow for extra fields (e.g. in training data) that may not
+            # be needed by the data objects, we ignore unknown fields
+            return element_type.from_json(raw_element, ignore_unknown_fields=True)
+        return raw_element
+
+    @staticmethod
+    def _to_element_partial(element_type: type) -> Callable:
+        return partial(DataStreamSourcePlugin._to_element_type, element_type)
+
+
+class FilePluginBase(DataStreamSourcePlugin):
```

Comment:
My expectation is that this will likely be very useful for derived implementations as well, especially those that are fetching files from external sources like S3